### PR TITLE
Batch revisted

### PR DESCRIPTION
**Pull request by markusdregi**
**Task**
_Change `BatchSimulator.start` to take named control variable values._

**Approach**
_Pass them as a dict instead of a list._

**Pre un-WIP checklist**
- [x] Statoil tests pass locally

### CODE REVIEW

**Comment by markusdregi** [comment]
@joakim-hove Please have a look to see that you agree regarding the API changes.

Please note that the formatting went kind of a wall. Will fix that in a commit no, so no need for formatting comments yet ;)

**Comment by joakim-hove** [comment]
First of all: I assume this implies that seba is now ready to pass named parameters to the simulator?

File: python/python/res/simulator/batch_simulator.py [hunk 1]
```diff
@@ -107,20 +120,25 @@ def start(self, case_name, controls):
...
-               node = EnkfNode(config_node)
-               ext_node = node.as_ext_param( )
-               ext_node.set_vector(values)
-               node.save(fs, node_id)
+            if set(control_dict.keys()) != set(self.controls_format.keys()):
+                err_msg = "Mismatch between initialized and provided control names."
+                raise KeyError(err_msg)
+
+            for control_name, control in control_dict.iteritems():
+                if set(control.keys()) != set(self.controls_format[control_name]):
```
Comment (joakim-hove) [inline python/python/res/simulator/batch_simulator.py]: I am not too happy about this duplication of logic/configuration of keys, the `ext_param` implementation is the authorative source regarding these keys. The 
```Python
for var_name, value in control.iteritems():     
    ext_node[var_name] = value
```
will already raise a `KeyError` for invalid key. Could in addition have:
```Python
if len(ext_config) != len(control):
   raise ValueError("Not all keys set")
```
alternatively a convenience method could be added here: https://github.com/Statoil/libres/blob/master/python/python/res/enkf/config/ext_param_config.py

**Comment by joakim-hove** [comment]
Well tested - apart from comment about duplicated validation logic this is ready to be merged (when Seba is!!)

**Comment by markusdregi** [comment]
@joakim-hove: Yes, _Seba_ is somewhat ready for this. It's not perfect, but it is already extracting the necessary information from its tags. So we'll just use that for now and improve upon it with something more explicit in the future.

File: python/python/res/simulator/batch_simulator.py (hunk 1 above)
Comment (markusdregi) [reply python/python/res/simulator/batch_simulator.py]: I completely agree. I've fixed it now.

**Comment by markusdregi** [comment]
Alright, I consider this one to be done. But I'll leave it unmerged for a little while to create relevant patches.

**Comment by markusdregi** [comment]
The footprint of the last commit is relatively large and its solely a refactoring commit. In other words, the correctness of this PR is completely mute to whether we include the last commit or not. To summarize, it does the following:

- reformat code, in particular the dictionaries defining the batches. This is done as a join of my own personal taste and pylint recommendation, with the later taking precedence.
- Split one large test function into several smaller ones.
- Extract the logic that sets up the controls in `BatchSimulator.start` into a private method.

I leave it up to you to decide whether it should be included or not.

**Comment by joakim-hove** [comment]
Details about formatting aside - is this ready for the green button? I guess there is some ordering requirement with other PR's?

**Comment by markusdregi** [comment]
Yes, this one should go first, but will immediately break master _Seba_. The PR there is ready as well. In _Everest_ there should be no changes necessary. I'm running its tests with the two PR's as we speak to be sure.

I was thinking that we get both PR's reviewed and accepted first. Then, we merge them one by one as the tests are passing.

File: python/python/res/simulator/batch_simulator.py [hunk 2]
```diff
@@ -50,27 +50,48 @@ def __init__(self, res_config, controls, results):
             raise ValueError("The first argument must be valid ResConfig instance")
 
         self.res_config = res_config
-        self.ert = EnKFMain( self.res_config )
-        self.control_keys = []
-        self.result_keys = []
+        self.ert = EnKFMain(self.res_config)
+        self.control_keys = tuple(controls.keys())
```
Comment (joakim-hove) [inline python/python/res/simulator/batch_simulator.py:54]: Constification with `tuple` was a nice move

File: python/python/res/simulator/batch_simulator.py [hunk 3]
```diff
@@ -50,27 +50,48 @@ def __init__(self, res_config, controls, results):
...
+        for sim_id, (geo_id, controls)  in enumerate(case):
+            assert isinstance(geo_id, int)
+
+            node_id = NodeId(0, sim_id)
+            if set(controls.keys()) != set(self.control_keys):
+                err_msg = "Mismatch between initialized and provided control names."
+                raise KeyError(err_msg)
+
+            for control_name, control in controls.iteritems():
+                ens_config = self.res_config.ensemble_config
```
Comment (joakim-hove) [inline python/python/res/simulator/batch_simulator.py:75]: I would prefer if the `ens_config` variable  was moved out of both the loops.

**Review by joakim-hove — APPROVED** [review]

File: python/python/res/simulator/batch_simulator.py (hunk 3 above)
Comment (markusdregi) [reply python/python/res/simulator/batch_simulator.py:75]: It is just a shorter name for the `ensemble_config` in question. In my opinion it is odd to define it before needed. How about skipping the shorter name as it is only used once and go for
```python
node = EnkfNode(self.res_config.ensemble_config[control_name])
```
78 characters long and good to go?

**Comment by joakim-hove** [comment]
No big deal. My point is just that if it is extracted outside, as I suggest, that will implicitly communicate that this variable is fully immutable.

**Pull request merged**